File: src/javascript/algorithms/array/medianSortedArrays/MedianSortedArrays.js
function medianSortedArrays(nums1, nums2) {
    if (nums1.length > nums2.length) return medianSortedArrays(nums2, nums1);
    const n1 = nums1.length, n2 = nums2.length;
    const total = n1 + n2;
    const half = Math.floor(total / 2);
    let left = 0, right = n1 - 1;
    while (true) {
        const i = Math.floor((left + right) / 2);
        const j = half - i - 2;
        const firstLeft = i >= 0 ? nums1[i] : -Infinity;
        const firstRight = (i + 1) < n1 ? nums1[i + 1] : Infinity;
        const secondLeft = j >= 0 ? nums2[j] : -Infinity;
        const secondRight = (j + 1) < n2 ? nums2[j + 1] : Infinity;
        if (firstLeft <= secondRight && secondLeft <= firstRight) {
            if (total % 2) {
                return Math.min(firstRight, secondRight);
            }
            return (Math.max(firstLeft, secondLeft) + Math.min(firstRight, secondRight)) / 2;
        } else if (firstLeft > secondRight) {
            right = i - 1;
        } else {
            left = i + 1;
        }
    }
}

// Example calls
console.log("Median (Even total):", medianSortedArrays([1, 3], [2, 4])); // Expected: 2.5
console.log("Median (Odd total):", medianSortedArrays([1, 2], [3, 4, 5])); // Expected: 3.0
console.log("Median (One empty):", medianSortedArrays([], [1])); // Expected: 1.0
try {
    console.log("Median (Both empty):", medianSortedArrays([], [])); // Expected: Exception or error
} catch (e) {
    console.log("Exception:", e.message);
}
console.log("Median (Negatives):", medianSortedArrays([-5, 3, 6], [-2, 4, 10])); // Expected: 3.5;

module.exports = { medianSortedArrays };

File: src/python/algorithms/array/medianSortedArrays/MedianSortedArrays.py
def median_sorted_arrays(nums1, nums2):
    # Ensure nums1 is the smaller array
    if len(nums1) > len(nums2):
        return median_sorted_arrays(nums2, nums1)
    n1, n2 = len(nums1), len(nums2)
    total = n1 + n2
    half = total // 2
    left, right = 0, n1 - 1
    while True:
        i = (left + right) // 2 if n1 > 0 else -1
        j = half - i - 2
        first_left = nums1[i] if i >= 0 else float('-inf')
        first_right = nums1[i+1] if (i+1) < n1 else float('inf')
        second_left = nums2[j] if j >= 0 else float('-inf')
        second_right = nums2[j+1] if (j+1) < n2 else float('inf')
        if first_left <= second_right and second_left <= first_right:
            if total % 2:
                return min(first_right, second_right)
            return (max(first_left, second_left) + min(first_right, second_right)) / 2
        elif first_left > second_right:
            right = i - 1
        else:
            left = i + 1

if __name__ == "__main__":
    # Test case 1: Even total length
    nums1 = [1, 3]
    nums2 = [2, 4]
    print("Median (Even total):", median_sorted_arrays(nums1, nums2))  # Expected: 2.5

    # Test case 2: Odd total length
    nums3 = [1, 2]
    nums4 = [3, 4, 5]
    print("Median (Odd total):", median_sorted_arrays(nums3, nums4))  # Expected: 3.0

    # Test case 3: One array empty
    nums5 = []
    nums6 = [1]
    print("Median (One empty):", median_sorted_arrays(nums5, nums6))  # Expected: 1.0

    # Test case 4: Both arrays empty
    nums7 = []
    nums8 = []
    try:
        print("Median (Both empty):", median_sorted_arrays(nums7, nums8))  # Expected: Exception or error
    except Exception as e:
        print("Exception:", e)

    # Test case 5: Arrays with negative numbers
    nums9 = [-5, 3, 6]
    nums10 = [-2, 4, 10]
    print("Median (Negatives):", median_sorted_arrays(nums9, nums10))  # Expected: 3.5
